_drafts/firmware-update-architecture.md

```diff
@@ -0,0 +1,834 @@
+identifies what type of image it is, how large it is, what version it is, ...
+etc.
+
+That metadata must be:
+1. Easy to find and parse for an external program
+2. Accessible from within the firmware itself
+
+My favorite approach is to add a struct at the beginning of the image to serve
+as a header. Here's what it looks like:
+
+```c
+// image.h
+typedef struct __attribute__((packed)) {
+    uint16_t image_magic;
+    uint16_t image_hdr_version;
+    uint32_t crc;
+    uint32_t data_size;
+    uint8_t image_type;
+    uint8_t version_major;
+    uint8_t version_minor;
+    uint8_t version_patch;
+    uint32_t vector_addr;
+    uint32_t reserved;
+    char git_sha[8];
+} image_hdr_t;
+```
+
+The contents of that header will vary per application, but you must include:
+1. A way to indicate that the header is valid (here we used a constant in
+`image_magic`)
+2. A version for the header, and
+3. The address of the start of your vector table.
+
+Each of our images must fill in its own header. For example, here it is for my
+Application:
+
+```c
+// app.c
+image_hdr_t image_hdr __attribute__((section(".image_hdr"))) = {
+    .image_magic = IMAGE_MAGIC,
+    .image_hdr_version = IMAGE_VERSION_CURRENT,
+    .image_type = IMAGE_TYPE_APP,
+    .version_major = 1,
+    .version_minor = 0,
+    .version_patch = 1,
+    .vector_addr = (uint32_t)&vector_table,
+    // GIT_SHA is generated by Makefile
+    .git_sha = GIT_SHA,
+};
+```
+
+You will notice the `section` attribute, which is used to place the variable in
+a specific section of our firmware. This is paired with some code in our linker
+script which declares that sections and puts it at the very start of flash:
+
+```
+// app.ld
+SECTIONS
+{
+    .image_hdr : {
+        KEEP (*(.image_hdr))
+    } > slot2rom
+
+    /* Vector table must be 128-bit aligned */
+    . = ALIGN(128);
+
+    .text : {
+        *(.vectors)    /* Vector table */
+        *(.text*)    /* Program code */
+        . = ALIGN(4);
+        *(.rodata*)    /* Read-only data */
+        . = ALIGN(4);
+    } >slot2rom
+
+    ...
+};
+```
+
+Some data can simply be hardcoded into the header. In my case, this includes the
+`image_magic`, the semantic version, the image type, the header version, and the
+vector table address. Other bits must be read from the environment and injected
+as `DEFINES` by our Makefile. For example, this is how we get the `GIT_SHA`:
+
+```makefile
+# Makefile
+GIT_SHA := \"$(shell $(GIT) rev-parse --short HEAD)\"
+DEFINES += GIT_SHA=$(GIT_SHA)
+CFLAGS += $(foreach d,$(DEFINES),-D$(d))
+```
+
+The `GIT_SHA` macro can then be called in our code to get the short (7-char)
+hash as a string.
+
+Some header information can only be calculated *after* the firmware has been
+compiled and linked. For example, the length of the binary, its CRC, or its
+cryptographic signature. To add it to the header, we must edit the binary
+directly.
+
+Typically, I would write a Python script to do this but in a pinch you can use
```

Comment:
should be able to replace with the python script at https://github.com/memfault/interrupt/commit/fc1065bc555b6fdb6f9c3340841b6368b4287e7d